pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -280,10 +281,13 @@ def restore_optimizers(self) -> None:
             # move optimizer to GPU 1 weight at a time
             # avoids OOM
             if self.trainer.root_gpu is not None:
-                for state in optimizer.state.values():
-                    for k, v in state.items():
-                        if isinstance(v, torch.Tensor):
-                            state[k] = v.cuda(self.trainer.root_gpu)
+                for param, state in optimizer.state.items():
+                    if isinstance(state, collections.Mapping):
+                        for k, v in state.items():
+                            if isinstance(v, torch.Tensor):
+                                state[k] = v.cuda(self.trainer.root_gpu)
+                    elif isinstance(state, torch.Tensor):
+                        optimizer.state[param] = state.cuda(self.trainer.root_gpu)
```

Comment:
thanks for working on this!
This logic doesn't need to be in the checkpoint connector anymore. It can go here: 
https://github.com/PyTorchLightning/pytorch-lightning/blob/8ddf9f996d4a98e7be5fc591b7734b6f78313771/pytorch_lightning/strategies/strategy.py#L320-L323

then we can delete lines 280-290 from here

Inside the strategy class, you can replace `self.trainer.root_gpu` with `self.root_device`

this is what it would look like after:
```py
if self.root_device.type == "cuda":
    for param, state in optimizer.state.items():
        if isinstance(state, collections.Mapping):
            for k, v in state.items():
                if isinstance(v, torch.Tensor):
                    state[k] = v.to(self.root_device)
        elif isinstance(state, torch.Tensor):
            optimizer.state[param] = state.to(self.root_device)
```

Reply:
Actually I wonder if we could directly use this utility: https://github.com/PyTorchLightning/pytorch-lightning/blob/8ddf9f996d4a98e7be5fc591b7734b6f78313771/pytorch_lightning/strategies/strategy.py#L143-L149

Reply:
> thanks for working on this! This logic doesn't need to be in the checkpoint connector anymore. It can go here:
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/8ddf9f996d4a98e7be5fc591b7734b6f78313771/pytorch_lightning/strategies/strategy.py#L320-L323
> 
> then we can delete lines 280-290 from here
> 
> Inside the strategy class, you can replace `self.trainer.root_gpu` with `self.root_device`
> 
> this is what it would look like after:
> 
> ```python
> if self.root_device.type == "cuda":
>     for param, state in optimizer.state.items():
>         if isinstance(state, collections.Mapping):
>             for k, v in state.items():
>                 if isinstance(v, torch.Tensor):
>                     state[k] = v.to(self.root_device)
>         elif isinstance(state, torch.Tensor):
>             optimizer.state[param] = state.to(self.root_device)
> ```
> 
> fyi @four4fish as this will simpliy device properties further by removing a dependency on `root_gpu`

I'm wondering which is the better way:
1) create a new PR for the code refactoring, and keep this PR solely for the bug fixing purpose
2) fix the bug with the code refactoring

Please tell me which way is preferred as a better practice for Pytorch Lightning. Thanks.

Reply:
Keep this PR solely for the bugfix. I created #11758 for the refactoring